src/unittest/python/core/attributes/__init__.py

```diff
@@ -0,0 +1,13 @@
+# Copyright 2018 JanusGraph Python Authors
```

Comment:
Yes, Should I remove the folder is it? I'm kinda out of ideas what sort of tests should be constructed here?

Reply:
You'll probably need integration tests that actually query a running JanusGraph database. You could maybe use [testcontainers-python](https://github.com/testcontainers/testcontainers-python) for that with for example [this docker image](https://hub.docker.com/r/florianhockmann/janusgraph-testing) which already has the Graph of the Gods loaded. (I haven't used testcontainers-python yet though, only testcontainers-dotnet which I'm using for JanusGraph.Net.)

Reply:
I've been thinking the same since long, to develop Integration tests. Thanks for pointing in right direction. I had issue with using testcontainer-python, in sense that I didn't find any way to pull images from dockerhub using testcontainers-python. I then googled and switched to [Docker SDK for Python](https://docker-py.readthedocs.io/en/stable/index.html). 

I'm able to pull your image for janusgraph-testing, and able to start container using Python also. But I'm facing errors now connecting to the docker image. Since I've very less experience with Docker, so this error has been creating problems. So any help here?

Run docker container:

```
container = client.containers.run("florianhockmann/janusgraph-testing:0.3.0", detach=True,
                                      hostname="0.0.0.0", ports={8182: ("0.0.0.0", 8182)})
```

I then do `docker ps` from command prompt, and I see following:

![image](https://user-images.githubusercontent.com/1013906/51196572-6de14c00-1915-11e9-951d-b17e02ce7f09.png)

As you can see, the docker image is up and running, and is connected at 0.0.0.0 (8182).

Now, when I connect, why I'm I getting OSError? Whereas I can use the same code to connect to another instance of Gremlin server in Non Docker mode.

The following fails:

```
client = JanusGraphClient().connect(host="0.0.0.0", port="8182", traversal_source="g").get_connection()
g = Graph().traversal().withRemote(client)
```

Error:

```
Traceback (most recent call last):
  File "C:/Projects/JanusGraph/janusgraph-python/src/main/scripts/janusgraph-python", line 47, in <module>
    client = JanusGraphClient().connect(host="0.0.0.0", port="8182", traversal_source="g").get_connection()
  File "C:\Users\IBM_ADMIN\Anaconda3\envs\janusgraph-python\lib\site-packages\janusgraph_python\driver\ClientBuilder.py", line 74, in connect
    graphson_writer=graphson_writer)
  File "C:\Users\IBM_ADMIN\Anaconda3\envs\janusgraph-python\lib\site-packages\gremlin_python\driver\driver_remote_connection.py", line 45, in __init__
    password=password)
  File "C:\Users\IBM_ADMIN\Anaconda3\envs\janusgraph-python\lib\site-packages\gremlin_python\driver\client.py", line 76, in __init__
    self._fill_pool()
  File "C:\Users\IBM_ADMIN\Anaconda3\envs\janusgraph-python\lib\site-packages\gremlin_python\driver\client.py", line 88, in _fill_pool
    conn = self._get_connection()
  File "C:\Users\IBM_ADMIN\Anaconda3\envs\janusgraph-python\lib\site-packages\gremlin_python\driver\client.py", line 101, in _get_connection
    self._transport_factory, self._executor, self._pool)
  File "C:\Users\IBM_ADMIN\Anaconda3\envs\janusgraph-python\lib\site-packages\gremlin_python\driver\connection.py", line 40, in __init__
    self.connect()
  File "C:\Users\IBM_ADMIN\Anaconda3\envs\janusgraph-python\lib\site-packages\gremlin_python\driver\connection.py", line 46, in connect
    self._transport.connect(self._url)
  File "C:\Users\IBM_ADMIN\Anaconda3\envs\janusgraph-python\lib\site-packages\gremlin_python\driver\tornado\transport.py", line 33, in connect
    lambda: websocket.websocket_connect(url))
  File "C:\Users\IBM_ADMIN\Anaconda3\envs\janusgraph-python\lib\site-packages\tornado\ioloop.py", line 458, in run_sync
    return future_cell[0].result()
  File "C:\Users\IBM_ADMIN\Anaconda3\envs\janusgraph-python\lib\site-packages\tornado\concurrent.py", line 238, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 4, in raise_exc_info
OSError: [Errno 10049] Unknown error
```

Reply:
Oh thanks for that. I'm able to setup pull your docker image, start it from my Python script, and able to query it also now. 

As for host, well I checked, if I run from docker on windows (<=7), then the IP is 192.168.99.100, whereas that is localhost if we run from Unix or Mac. Now such variability is a problem for dynamically retriving IP and passing it to client.

So, I created a subprocess, which runs following shell script and retrive IP from shell command. The command I run is `docker-machine ip` and the output of that is used as `host` for client.

I craeted #19 to track the development of integraton tests. Any idea on how and what the tests should contain as this is first time I'm developing Integration tests.

Reply:
Also, realized that now. Should I do changes to move `port` to integer from string?

Reply:
> As for host, well I checked, if I run from docker on windows (<=7), then the IP is 192.168.99.100, whereas that is localhost if we run from Unix or Mac.

You just need to [expose the port](https://docs.docker.com/engine/reference/run/#expose-incoming-ports) to make it accessible on `localhost`. The command for that is `docker run -p 8182:8182 ...`. I don't know how this works with this SDK you're using, although the code you showed earlier seemed to do exactly that.

> I craeted #19 to track the development of integraton tests. Any idea on how and what the tests should contain as this is first time I'm developing Integration tests.

I think that a library should have tests for all functionality available for users that is non-trivial. Here, I would at least have one test per text predicate for example. So, one test that uses `textContains()` for example and then does a count and assert that the count is as expected, and so on for the other predicates.
My opinion is that we shouldn't release a library where we don't have these tests as we can't have the confidence that it actually works without the tests. If you don't want to do them in this PR to limit the scope, then it should be done in a subsequent PR before the first release. Although I personally would prefer if the tests would be included in this PR as it makes the review of course easier if we know from the tests that the code is functionally correct.

Reply:
Yes, I agree with you there. Every functionality needs to have a test corresponding to it, atleast as many as possible, and not having tests for whole `attributes` sub package makes it really dangerous.

Hence, though I've created that #19 , but I'm targetting to close that within scope of this PR. To be honest, you can already see tests for Text Attributes in, and those pass successfully also. The latest build fixes #19 with respect to Text Attributes. Guess others tests needed would be `Geo Predicates` I guess. Is there anything else except `Text Predicates & Geo Predicates`? Or Do we also need tests for `ClientBuilder`?